### PR TITLE
Create Article : Data-Structures-Trie.md

### PR DESCRIPTION
**Pull request by ds-249**

### CODE REVIEW

File: Data-Structures-Trie.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
+# Trie
+
+## Introduction to Trie
+
+The word trie is an inflix of the word "re**trie**val", because the trie can find a single word in a dictionary with only a prefix of the word.
+Trie is an efficient data retrieval data structure, using trie, search complexities can be brought to an optimal limit, i.e. length of the string.
+It is a multi-way tree structure useful for storing strings over an alphabet.
```
Comment (alayek) [inline Data-Structures-Trie.md]:
```diff
- It is a multi-way tree structure useful for storing strings over an alphabet.
+ It is a multi-way tree structure, useful for storing strings over an alphabet, when we are storing them.
```

File: Data-Structures-Trie.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+It has been used to store large dictionaries (do not confuse this with the dictionaries of python) of any language, say English, words in spell-checking programs.
```
Comment (alayek) [inline Data-Structures-Trie.md]: > It has been used to store large dictionaries (do not confuse this with the dictionaries of python) of any language, say English, words in spell-checking programs.

@atjonathan any suggestions?

File: Data-Structures-Trie.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+However, the penalty on tries is the storage requirement.
+
+## What is a trie?
+
+A trie is a tree like data structure which stores strings, and helps you find the data associated with that string using the prefix of the string.
```
Comment (alayek) [inline Data-Structures-Trie.md]: Can you refer to some image from external source?

File: Data-Structures-Trie.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+The word trie is an inflix of the word "re**trie**val", because the trie can find a single word in a dictionary with only a prefix of the word.
+Trie is an efficient data retrieval data structure, using trie, search complexities can be brought to an optimal limit, i.e. length of the string.
+It is a multi-way tree structure useful for storing strings over an alphabet.
+It has been used to store large dictionaries (do not confuse this with the dictionaries of python) of any language, say English, words in spell-checking programs.
+However, the penalty on tries is the storage requirement.
+
+## What is a trie?
+
+A trie is a tree like data structure which stores strings, and helps you find the data associated with that string using the prefix of the string. 
+For example, say you plan on building a dictionary to store strings along with their meanings. You must be wondering why can't I simply use a hash table, to get the information. Yes, you obviously can get information using a hash table, but the worst time complexity for fetching data from a hash table is `O(n)`, where `n` is the number of strings stored.
```
Comment (alayek) [inline Data-Structures-Trie.md]: Well, technically it is `O(1)`, however, with tries, further searches work better when your string is growing by size (in an auto-suggest program, for example).

File: Data-Structures-Trie.md (hunk 3 above)
Comment (ds-249) [inline Data-Structures-Trie.md]: @alayek I have added one in the implementation section, should I move it up here?

File: Data-Structures-Trie.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+For example, say you plan on building a dictionary to store strings along with their meanings. You must be wondering why can't I simply use a hash table, to get the information. Yes, you obviously can get information using a hash table, but the worst time complexity for fetching data from a hash table is `O(n)`, where `n` is the number of strings stored. 
+But when you use a trie for fetching data, the time complexity to get data is `O(w)`, where `w` is the length of the string.
+
+## How to implement a trie?
+
+A trie typically, looks something like this,
+
+![Trie](https://community.topcoder.com/i/education/alg_tries.png)
+
+The above image stores, assoc, algo, all, also, tree, trie.
```
Comment (alayek) [inline Data-Structures-Trie.md]: I am not sure what this means.

File: Data-Structures-Trie.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+class TrieNode:
+	def __init__(self):
+		self.edges = [None]*(ALPHABET_SIZE) # Each index respective to each character.
+		self.meaning = None # Meaning of the word.
+		self.ends_here = False # Tells us if the word ends here.
+```
+As you can see, edges are 26 in length, each index referring to each character in the alphabet. 'A' corresponding to 0, 'B' to 1, 'C' to 2 ... 'Z' to 25th index. If the character you are looking for is pointing to `None`, that implies the word is not there in the trie.
+
+A typical Trie should implement at least these two functions:
```
Comment (alayek) [inline Data-Structures-Trie.md]: Wrap the function names in inline code blocks.

File: Data-Structures-Trie.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+			return self.meaning
+		ch = word[0]
+		index = ord(ch) - ord('a')
+		if self.edges[index] == None:
+			return "Word doesn't exist in the Trie"
+		else:
+			return self.edges[index].get_meaning(word[1:])
+```
+
+:rocket: [Run Code](https://repl.it/CWaJ)
```
Comment (alayek) [inline Data-Structures-Trie.md]: Can you add some example in the REPL snippet with these functions?

Comment (ds-249) [inline Data-Structures-Trie.md]: I have added one.

Comment (alayek) [inline Data-Structures-Trie.md]: No, I mean, you use them in a main function for inserting, deleting, searching etc.

Comment (ds-249) [inline Data-Structures-Trie.md]: Updated it.

**Comment by atjonathan** [comment]
### **LGTM :+1:**

**Comment by Rafase282** [comment]
# LGTM :+1:

**Comment by atjonathan** [comment]
@alayek @koustuvsinha, please take a look :tada:

File: Data-Structures-Trie.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+The word trie is an inflix of the word "re**trie**val", because the trie can find a single word in a dictionary with only a prefix of the word.
+Trie is an efficient data retrieval data structure, using trie, search complexities can be brought to an optimal limit, i.e. length of the string.
+It is a multi-way tree structure useful for storing strings over an alphabet, when we are storing them.
+It has been used to store large dictionaries of English, say, words in spell-checking programs.
+However, the penalty on tries is the storage requirement.
+
+## What is a trie?
+
+A trie is a tree like data structure which stores strings, and helps you find the data associated with that string using the prefix of the string.
+For example, say you plan on building a dictionary to store strings along with their meanings. You must be wondering why can't I simply use a hash table, to get the information.
```
Comment (koustuvsinha) [inline Data-Structures-Trie.md]: Reference hash table : the existing article `Hash-Tables-and-Hashing Functions.md`

**Comment by koustuvsinha** [comment]
Just add the reference, otherwise

# LGTM :+1:

File: Data-Structures-Trie.md [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+Yes, you obviously can get information using a hash table, but, the [hash tables](Hash-Tables-And-Hashing-Functions.md) can only find data where the string exactly matches the one we've added. But trie will give us the capability to find strings with common prefixes, a missing character etc in lesser time, in comparison to a hash table.
```
Comment (Rafase282) [inline Data-Structures-Trie.md]:
```diff
-[hash tables](Hash-Tables-And-Hashing-Functions.md)
+[hash tables](Hash-Tables-And-Hashing-Functions)
```

**Comment by atjonathan** [comment]
:tada: Thanks @ds-249 :tada:

**Pull request merged**